qolmat/analysis/holes_characterization.py

```diff
@@ -88,7 +101,545 @@ def test(self, df: pd.DataFrame) -> float:
+
+        if self.exact_p_value:
+            self.process_permutation = self._parallel_process_permutation_exact
+        self.process_permutation = self._parallel_process_permutation
+
+    def _encode_dataframe(self, df: pd.DataFrame) -> np.ndarray:
+        """
+        Encodes the DataFrame by converting numeric columns to a numpy array
+        and applying one-hot encoding to non-numeric columns.
+
+        Parameters:
+        -----------
+        df : pd.DataFrame
+            The DataFrame to be encoded.
+
+        Returns:
+        -------
+        np.ndarray
+            The encoded DataFrame as a numpy ndarray, with numeric data concatenated
+            with one-hot encoded categorical and boolean data.
+        """
+        if not self.encoder:
+            self.encoder = OneHotEncoder(
+                cols=df.select_dtypes(include=["object", "bool"]).columns,
+                return_df=False,
+                handle_missing='return_nan'
+            )
+
+        return self.encoder.fit_transform(df)
+
+    def _pklm_preprocessing(self, X: Union[pd.DataFrame, np.ndarray]) -> np.ndarray:
+        """
+        Preprocesses the input DataFrame or ndarray for further processing.
+
+        Parameters:
+        -----------
+        X : Union[pd.DataFrame, np.ndarray]
+            The input data to be preprocessed. Can be a pandas DataFrame or a numpy ndarray.
+
+        Returns:
+        -------
+        np.ndarray
+            The preprocessed data as a numpy ndarray.
+
+        Raises:
+        -------
+        TypeNotHandled
+            If the DataFrame contains columns with data types that are not numeric, string, or boolean.
+        """
+        if isinstance(X, np.ndarray):
+            return X
+
+        check_pd_df_dtypes(
+            X,
+            [
+                pd.api.types.is_numeric_dtype,
+                pd.api.types.is_string_dtype,
+                pd.api.types.is_bool_dtype
+            ]
+        )
+        return self._encode_dataframe(X)
+
+    @staticmethod
+    def _get_max_draw(p: int) -> int:
+        """
+        Calculates the number of possible projections.
+
+        Parameters:
+        -----------
+        p : int
+            The number of columns of the input matrix.
+
+        Returns:
+        --------
+        int
+            The number of possible projections.
+        """
+        return p*(2**(p-1) - 1)
+
+    def _draw_features_and_target_indexes(self, X: np.ndarray) -> Tuple[List[int], int]:
+        """
+        Randomly selects features and a target from the dataframe.
+
+        Parameters:
+        -----------
+        X : np.ndarray
+            The input dataframe.
+
+        Returns:
+        --------
+        Tuple[np.ndarray, int]
+            Indices of selected features and the target.
+        """
+        _, p = X.shape
+        nb_features = self.rng.randint(1, p)
+        features_idx = self.rng.choice(range(p), size=nb_features, replace=False)
```

Comment:
with the fonction "choice", you can directly write : self.rng.choice(p, size=....) -> more efficient since you do not need to create a range